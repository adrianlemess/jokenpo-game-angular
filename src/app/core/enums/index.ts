export * from './end-game-result';
export * from './hand-type';
export * from './player-type';
