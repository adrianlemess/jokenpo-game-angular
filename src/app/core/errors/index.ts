export class StrategyError implements Error {
    name: string;
    message: string;
    stack?: string | undefined;

    constructor(name: string) {
        (this.name = name), (this.message = `Hand not implemented on ${this.name} `);
    }
}
