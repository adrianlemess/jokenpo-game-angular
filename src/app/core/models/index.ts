export * from './paper-hand.model';
export * from './rock-hand.model';
export * from './scissor-hand.model';
export * from './game-match.model';
export * from './player.model';
export * from './hand.model';
