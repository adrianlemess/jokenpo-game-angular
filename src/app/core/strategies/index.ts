export * from './paper.strategy';
export * from './rock.strategy';
export * from './scissor.strategy';
