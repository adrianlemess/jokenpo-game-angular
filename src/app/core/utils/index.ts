/**
 * Function to return a enum.
 * Got from: https://stackoverflow.com/questions/44230998/how-to-get-a-random-enum-in-typescript
 * I've decided to isolate this logic in a function to facilitate further tests
 */

export const randomEnum = <T>(anEnum: T): T[keyof T] => {
    const enumValues = Object.values(anEnum) as unknown as T[keyof T][];
    const randomIndex = Math.floor(Math.random() * enumValues.length);

    return enumValues[randomIndex];
};
