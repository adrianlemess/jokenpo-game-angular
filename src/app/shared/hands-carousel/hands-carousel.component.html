<div class="carousel container-fluid">
    <h1>{{ carouselTitle }}</h1>
    <h3>{{ carouselSubtitle }}</h3>

    <ngb-carousel #carousel>
        <ng-template ngbSlide *ngFor="let hand of hands" [id]="hand.name">
            <div class="wrapper carousel">
                <img [src]="hand.avatarImgUrl" class="image" [alt]="'Hand image' + hand.name" />
            </div>
        </ng-template>
    </ngb-carousel>
</div>
