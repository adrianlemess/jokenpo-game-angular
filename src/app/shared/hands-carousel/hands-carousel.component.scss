.carousel {
    margin: 32px auto 16px;
    text-align: center;
    & h3 {
        text-transform: capitalize;
    }
    &__caption {
        position: absolute;
        right: 15%;
        bottom: 1.25rem;
        left: 15%;
        padding-top: 1.25rem;
        padding-bottom: 1.25rem;
        color: #fff;
        text-align: center;
    }

    & img {
        object-fit: cover;
        max-width: 50%;
        height: auto;
        overflow: hidden;
    }
}
