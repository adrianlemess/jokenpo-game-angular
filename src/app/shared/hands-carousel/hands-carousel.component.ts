import { Component, Input } from '@angular/core';
import { NgbCarouselConfig } from '@ng-bootstrap/ng-bootstrap';
import { Hand } from '../../core/interfaces';

@Component({
    selector: 'app-hands-carousel',
    templateUrl: './hands-carousel.component.html',
    styleUrls: ['./hands-carousel.component.scss'],
})
export class HandsCarouselComponent {
    @Input()
    carouselTitle = '';

    @Input()
    carouselSubtitle = '';

    @Input()
    hands: Hand[];

    constructor(config: NgbCarouselConfig) {
        config.interval = 500;
        config.keyboard = false;
        config.pauseOnHover = false;
        config.showNavigationArrows = false;
        config.showNavigationIndicators = false;
    }
}
