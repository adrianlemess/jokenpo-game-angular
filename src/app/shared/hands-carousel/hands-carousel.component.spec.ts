import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { HandFactoryService } from '../../core/services/hand-factory.service';

import { HandsCarouselComponent } from './hands-carousel.component';

describe('HandsCarouselComponent', () => {
    let component: HandsCarouselComponent;
    let fixture: ComponentFixture<HandsCarouselComponent>;
    beforeEach(async () => {
        await TestBed.configureTestingModule({
            declarations: [HandsCarouselComponent],
            imports: [NgbModule],
        }).compileComponents();
    });
    beforeEach(() => {
        fixture = TestBed.createComponent(HandsCarouselComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
    });

    it('Should render with no hands and only carouselTitle and subtitle', () => {
        component.carouselTitle = 'Carousel title';
        component.carouselSubtitle = 'Carousel subtitle';
        component.hands = [];

        expect(fixture.nativeElement).toMatchSnapshot('Hands empty');
    });

    it('Should render with a list of hands', () => {
        const handsFactoryService = TestBed.inject(HandFactoryService);

        component.hands = handsFactoryService.getAllHands();
        component.carouselTitle = 'Carousel title';
        component.carouselSubtitle = 'Carousel subtitle';

        expect(fixture.nativeElement).toMatchSnapshot('Hands list');
    });
});
