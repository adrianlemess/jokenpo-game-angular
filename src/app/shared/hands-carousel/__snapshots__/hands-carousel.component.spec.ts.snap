// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HandsCarouselComponent Should render with a list of hands: Hands list 1`] = `
<div
  id="root1"
>
  <div
    class="carousel container-fluid"
  >
    <h1>
      
    </h1>
    <h3>
      
    </h3>
    <ngb-carousel
      aria-activedescendant="slide-"
      class="carousel slide"
      data-testid="carousel-hands"
      style="display: block;"
      tabindex="0"
    >
      <ol
        class="carousel-indicators visually-hidden"
        role="tablist"
      >
        
      </ol>
      <div
        class="carousel-inner"
      >
        
      </div>
      
      
    </ngb-carousel>
  </div>
</div>
`;

exports[`HandsCarouselComponent Should render with no hands and only carouselTitle and subtitle: Hands empty 1`] = `
<div
  id="root0"
>
  <div
    class="carousel container-fluid"
  >
    <h1>
      
    </h1>
    <h3>
      
    </h3>
    <ngb-carousel
      aria-activedescendant="slide-"
      class="carousel slide"
      data-testid="carousel-hands"
      style="display: block;"
      tabindex="0"
    >
      <ol
        class="carousel-indicators visually-hidden"
        role="tablist"
      >
        
      </ol>
      <div
        class="carousel-inner"
      >
        
      </div>
      
      
    </ngb-carousel>
  </div>
</div>
`;
